sound/soc/sof/intel/hda-dai.c

```diff
@@ -268,68 +253,83 @@ static int hda_dai_hw_params(struct snd_pcm_substream *substream,
 	return hda_link_dma_params(link_dev, &p_params);
 }
 
-static int hda_dai_prepare(struct snd_pcm_substream *substream,
-			   struct snd_soc_dai *dai)
+static int hda_dai_hw_params(struct snd_pcm_substream *substream,
+			     struct snd_pcm_hw_params *params,
+			     struct snd_soc_dai *dai)
+{
+	struct hdac_ext_stream *link_dev;
+	struct snd_soc_dapm_widget *w;
+	int stream_tag;
+	int ret;
+
+	ret = hda_dai_link_hw_params(substream, params);
+	if (ret < 0)
+		return ret;
+
+	link_dev = snd_soc_dai_get_dma_data(dai, substream);
+	if (!link_dev)
+		return -EINVAL;
+
+	stream_tag = hdac_stream(link_dev)->stream_tag;
+
+	if (substream->stream == SNDRV_PCM_STREAM_PLAYBACK)
+		w = dai->playback_widget;
+	else
+		w = dai->capture_widget;
+
```

Comment:
To simplify the code you can use:
`w = snd_soc_dai_get_widget(dai, substream->stream);`

Reply:
aye, thanks for the suggestion.

Reply:
done